### ISSUE TITLE
Incorrect annotation on top-level references causes function calls through references to not run

### ISSUE THREAD

**Issue by cpitclaudel** [FStarLang/FStar#1160]
I got bitten by the following while working on the compiler (repro by @nikswamy):

```
module Test
open FStar.All

val hook : ref (int -> ML unit)
let hook = alloc (fun _ -> ())

val do_stuff: unit -> ML unit
let do_stuff () = !hook 1
```

Extract this with `fstar test.fst --debug Test --debug_level LogTypes --codegen OCaml`

```
…
let do_stuff: Prims.unit -> Prims.unit =
  fun uu____162  -> let uu____163 = FStar_ST.op_Bang hook in () (* Never called *)
```

Here's the log:

```
Checked: val hook : FStar.ST.ref (uu___273061:Prims.int -> FStar.All.ML Prims.unit)

./test.fst(5,0-5,30): (Warning) Top-level let-bindings must be total; this term may have effects
Checked: let Test.hook =
  FStar.ST.alloc (fun uu___273366 -> ())
  <:
  FStar.ST.STATE
  (FStar.ST.mref (uu___273367:Prims.int -> Prims.Tot Prims.unit)
      (FStar.Heap.trivial_preorder (uu___273368:Prims.int -> FStar.All.ML Prims.unit)))
…
```

**Comment by cpitclaudel** [comment]
Thanks! I'm running into errors now though; I'll try to add more type annotations and see if that helps:

```
./u_boot_fsts/FStar.TypeChecker.Env.fst(1000,25-1000,50): (Error) Failed to resolve implicit argument of type 'Type' introduced in (*?u2*) _ uu___1351965 (see also ./u_boot_fsts/FStar.TypeChecker.Env.fst(1000,35-1000,50))
./u_boot_fsts/FStar.TypeChecker.Env.fst(1000,25-1000,50): (Error) Expected type "FStar.ST.ref (uu___1351947:FStar.TypeChecker.Env.env ->
    uu___1351948:FStar.TypeChecker.Env.binding -> FStar.All.ML Prims.unit)"; but "FStar.Util.mk_ref (fun uu___1351953 uu___1351954 -> ())
<:
FStar.All.ALL
(FStar.ST.ref (uu___1351955:(*?u1*) _ -> uu___1351956:(*?u2*) _ uu___1351955 -> Prims.Tot Prims.unit
    ))" has type "FStar.ST.ref (uu___1351951:(*?u1*) _ -> uu___1351952:(*?u2*) _ uu___1351951 -> Prims.Tot Prims.unit)"
2 errors were reported (see above)
```

**Comment by cpitclaudel** [comment]
That line looks like this:

```
let push_in_gamma_hook = BU.mk_ref (fun _ _ -> ())
```

And the interface says that:

```
val push_in_gamma_hook : ref<(env -> binding -> unit)>
```

**Comment by cpitclaudel** [comment]
Here's what the error message says after making types explicit

```
Expected type

"FStar.ST.ref (uu___1351947:FStar.TypeChecker.Env.env ->
    uu___1351948:FStar.TypeChecker.Env.binding -> FStar.All.ML Prims.unit)"

but

…

has type

"FStar.ST.ref (uu___1351951:FStar.TypeChecker.Env.env ->
    uu___1351952:FStar.TypeChecker.Env.binding -> Prims.Tot Prims.unit)"
```

So it infers Tot in one case and ML in the other one.  Why is it bad to use a Tot function where an ML one was expected?

**Comment by cpitclaudel** [comment]
The following trick seems to work:

```
let push_in_gamma_hook__ignore (_: env) (_: binding) : ML<unit> = ()
let push_in_gamma_hook = BU.mk_ref push_in_gamma_hook__ignore
```

**Comment by nikswamy** [comment]
Yes, indeed. I should have explained the problem and the fix here ... 

There was a bad bug in type inference that was allowing treating `(t -> Tot t)` as equal to `(t -> ML t)`, because certain equality problems were being turned mistakenly into subtyping problems.

This is now fixed. 

But, I noticed that this causes regressions in existing code especially when the `--eager_inference` flag is used. This flag explicitly disables certain forms of subtyping turning them into equalities. 

In particular, with `--eager_inference`   when you write `mk_ref #?u (fun () -> ())`, this unifies `?u` with `unit -> Tot unit` rather than saying only that `unit -> Tot unit <: ?u`. If what you actually wanted was to allocate a reference of type `ref (unit -> ML unit)`, then this is doomed. Previously, this was working because of the bug. If you remove the `--eager_inference` flag, then things should work out now.

So, as you'll see in my 0e0680bbd7fbd24b4b2959213ce41ec02a5a4667, I had to put in a few explicit promotions to the ML effect in the compiler to workaround this issue. See `FStar.Util.return_all`, which is basically a version of the trick you employed.
